lib/api/apiUtils/object/abortMultipartUpload.js

```diff
@@ -105,9 +112,47 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
                     }
                     cb();
                 });
+            }, () => next(null, mpuBucket, storedParts, destBucket, objectMD));
+        },
+        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objectMD, next) {
+            if (!objectMD || metadataValMPUparams.uploadId !== objectMD.uploadId) {
+                return next(null, mpuBucket, storedParts, destBucket, objectMD);
+            }
+            // In case there has been an error during cleanup after a complete MPU
+            // (e.g. failure to delete MPU MD in shadow bucket),
+            // we need to ensure that the MPU metadata is deleted.
+            log.info('Object has existing metadata, deleting them', {
+                method: 'abortMultipartUpload',
+                bucketName,
+                objectKey,
+            });
+            return metadata.deleteObjectMD(bucketName, objectKey, undefined, log, err => {
+                if (err) {
+                    log.error('error deleting object metadata', { error: err });
+                }
+                return next(err, mpuBucket, storedParts, destBucket, objectMD);
+            });
+        },
+        function deleteObjectData(mpuBucket, storedParts, destBucket, objectMD, next) {
+            if (!objectMD?.location) {
+                return next(null, mpuBucket, storedParts, destBucket);
+            }
+            // Filtering parts that has already been delete by the previous step
+            let partLocations = storedParts.map(item => item.value.partLocations);
+            partLocations = [].concat(...partLocations);
+            partLocations = new Set(partLocations.map(loc => loc.key));
+            const objectLocationLeft = objectMD.location.filter(loc => !partLocations.has(loc.key));
+
+            return async.eachLimit(objectLocationLeft, 5, (loc, cb) => {
+                data.delete(loc, log, err => {
+                    if (err) {
+                        log.fatal('delete object data failed', { err });
+                    }
+                    cb();
```

Comment:
We do not return any error in the callback yet log a fatal error: fatal is reserved to application crashes. Shall we log an error instead, and return the callback with the error?

Reply:
Not sure we want to do that, return an error would stop the async.eachLimit, not sure we want that (to create more data orphans..)
This is done this way just above in `deleteData` function